### PR TITLE
Implement HttpContent.writeTo() async API

### LINKED ISSUE #8790: Jetty-12 HttpContent should have an async API
@lachlan-roberts has recently been working on the content caching mechanism for Jetty-12. He's been receiving contradictory advice: I've been advocating that `HttpContent` should have an async API (maybe even just implement `Request.Processor`, but I think others have been advocating for keeping the blocking `getBuffer()` approach.

So I'm opening this issue to capture some of the arguments/discussions, but ultimately I think we need to have a meeting to pick a single direction and go with it.

The `ResourceService` in Jetty-12 has been greatly simplified, primarily by moving a lot of functionality out of special case handing into `ContentFactory` implementations and wrappers. Thus caching, cache invalidation etc. are all applied by wrapping / configuring / replacing the `ContentFactory`.   I think this is a great approach, but without an async API it means that only blocking behaviours can be moved into a `ContentFactory`.    Currently, we still have behaviours for directory listing, welcome files, precompressed content, range request etc. handled as conditional behaviours in the `ResourceService`.  If `HttpContent` was a `Processor` most of these complexities could be moved out to `ContentFactory` wrappers and only applied if actually needed/used.

A good example was that @sbordet has a use-case for handling welcome files with a redispatch through the context handlers:  the case where a welcome file might be index.php and to serve it it needs to be redispatched via the handler that will proxy a php to a different process.   With the current code, the `ContentFactory` will produce a `HttpContent` that represents the source code of the `index.php` file, probably even blocking for a little while in the constructor to load a `ByteBufffer` with the content that will never be used and creating `PreEncodedHttpFields` that are not relevant!  To handle this case, we must add a special welcome file mode to the `WelcomeFactory`, `WelcomeAction` and to the `ResourceService` itself.  This adds complexity even if a deployment doesn't need this feature.

If instead, the `HttpContent` was a `Request.Processor`, then a `WelcomeFileContentFactory` wrapper could replace any directory `HttpContent`s it receives from it's nested factory with a processor that would generate the contents of the directory.  This might be a redirect, generate a xhtml listing, or it may be a fully async redispatch through the handler tree, ultimately talking to a PHP server in another process.    Better yet, this could be done before the caching content factory, so the welcome file content could be cached.

Better yet, new ways of handling directories and/or welcome files can be added simply by adding new `ContentFactory` wrappers rather than adding more ifs and buts into the already complex `ResourceService`.

As far as I can tell, the counter argument is that this is "optimising the slow path".    I don't think that is the case because the async API is perfectly capable of being used to implement the simple direct behaviour that we currently have which  is (simplified):
```java
    public void doGet(Request request, Response response, Callback callback, HttpContent content) throws Exception
    {
            if (content.getResource().isDirectory())
            {
                sendWelcome(content, pathInContext, endsWithSlash, request, response, callback);
                return;
            }

            writeHttpContent(request, response, callback, content, reqRanges);
    }

    protected void sendWelcome(HttpContent content, String pathInContext, boolean endsWithSlash, Request request, Response response, Callback callback) throws Exception
    {
        // process optional Welcome behaviors
        if (welcome(request, response, callback))
            return;
        sendDirectory(request, response, content, callback, pathInContext);
    }

    protected void writeHttpContent(Request request, Response response, Callback callback, HttpContent content)
    {
        try
        {
            ByteBuffer buffer = content.getBuffer();
            if (buffer != null)
                response.write(true, buffer, callback);
            else
                new ContentWriterIteratingCallback(content, response, callback).iterate();
        }
        catch (Throwable x)
        {
            callback.failed(x);
        }
    }
```

I think this should just become:

```java
    public void doGet(Request request, Response response, Callback callback, HttpContent content) throws Exception
    {
        httpContent.process(request, response, callback);
    }
```

and by the miracle of object oriented programming, the content would serve itself with the best technique available.  If no welcome files are configured, then there is no welcome file handling code in the hot path. If welcome files are configured the content provided for a directory resource will either already be the welcome file, or be a processor that can run the welcome file logic.

If we wanted to have a simple cache that blocks loading the content into a buffer then we could had the implementation would end up something like:

```java
    public class CachedHttpContent extends HttpContent.Wrapper
    {
        private final ByteBuffer _content;

        public CachedHttpContent(HttpContent content)
        {
            _content = load(content);
        }
    
        public void process(Request request, Response response, Callback callback)
        {
            response.write(true, buffer, callback);
        }
    }
```

If that is the style of cache we want to have, there is no significant cost to this API approach, as it simply replaces if statements with virtual method call.  There may be some extra complexity in loading the wrapped HttpContent into a ByteBuffer, as we will need a fake request,response,callback, but that can be done in a utility class and will be generally useful anyway.    This is not "optimizing" the slow path.  This is generalising the hot path.

But by organising the API in this way, we can:
 + further simplify the minimal resource service, which now doesn't care about caching and we can do the same with welcome files etc.
 + decide to lazily load cached buffer, either keeping a list of calls to processes to complete, or forwarding them to the wrapped processor, until the content is loaded.
 + do arbitrary redispatching of content without any extra complications in the `ResourceService`
 + Cache directory and welcome file contents.  Perhaps even using a cache (third party?) that actually looks at headers such a `Vary` and `Cache-Control`
 + allow for arbitrary more complex behaviours in future.
        
In fact, we could almost go even further, and have the `HttpContent` as a `Request.Processor` be returned from the `handle` method of the `ResourceHandler`!?!?

In short, in Jetty-12 it is the `Request.Processor` API that we use to serve content. We wrap the processor to add behaviours like compression. We should use this same approach for static as well as dynamic content.

### PR DESCRIPTION
**Pull request by lorban**
Implementation of the `HttpContent.writeTo()` async API.

Fixes #8790

### CODE REVIEW

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/HttpContent.java
Comment (gregw) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/HttpContent.java]: Ah I see why we can't have request/response here!

But can we have variants that also take a ByteBufferPool.Sized?
We can't pass in a pool via the HttpContentFactory, as pools are possibly per connector, which means per request.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java]: maybe
```suggestion
        content.writeTo(response, 
            new ByteBufferPool.Sized(
                request.getComponents().getByteBufferPool(),
                request.getConnectionMetaData().getHttpConfiguration().getOutputBufferSize(),
                request.getConnectionMetaData().getHttpConfiguration().isUseOutputDirectByteBuffers()
            ),
            callback);
```

Better yet, add a `request.getComponents().getSizedByteBufferPool()` that could return a pool per connection rather than making the wrapper every time?

**Review by gregw — CHANGES_REQUESTED** [review]

**Comment by lorban** [comment]
@gregw There are two things I take from you previous comment:

 - `org.eclipse.jetty.server.Components` should expose `getSizedByteBufferPool()` _instead of_ `getByteBufferPool()` as its implementations should always have a default size. `IOResources` too should be modified to take a sized buffer pool where it expects the pool + size + directness triplet. This refactoring should be done independently in another PR.

 - Making `HttpContent.writeTo()` take a buffer pool as one of its arguments looks terrible from an API perspective: there is no fundamental reason why writing would require a buffer pool, and most implementations don't: only `ResourceHttpContent` needs one.
In my thinking, the sink fundamentally is linked to the channel that does hold a sized pool so we should just expose that by making `writeTo()` take a subinterface of `Content.Sink`. This is what I did and it doesn't look too bad, and definitely better than `writeTo()` taking a sized pool which I tried too.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java]: I think just have the one method that returns a Sized pool (but with javadoc saying what the sizes are).

```suggestion
    ByteBufferPool.Sized getByteBufferPool();
```

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/HttpContent.java
Comment (gregw) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/HttpContent.java]: I'm a big -1 on this interface.   Firstly associating a pool with a sink has nothing really to do with `HttpContent`. Any `Content.Sink` might ultimately have a buffer associated with it and then be wrapped with something like `BufferedContentSink`, which then has it's own buffers

let me ponder a better solution.....

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java]: I've changed my mind... we do need new methods because we have different sizes for input and output:

```suggestion
    ByteBufferPool getByteBufferPool();

    ByteBufferPool.Sized getInputByteBufferPool();
    ByteBufferPool.Sized getOutputByteBufferPool();
```

**Review by gregw — DISMISSED** [review]

**Comment by gregw** [comment]
@lorban I have created #12046 that improves the use of Sized ByteBufferPools.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java]: See #12046 which includes an implementation of this and also shows that using it can greatly simplify code

**Comment by lorban** [comment]
@gregw I like https://github.com/jetty/jetty.project/pull/12046 but it's not solving where should `ResourceHttpContent.writeTo()` should find its buffer pool.

Let me try something...

**Comment by lorban** [comment]
Here is a summary of the status of the `HttpContent` overhaul:

in progress:

 - add sized `ByteBufferPool`s in `Components` (https://github.com/jetty/jetty.project/pull/12046)
 - add `HttpContent.writeTo()` (the core of this PR)
 - no ByteBufferPool in the signature of `HttpContent.writeTo()` as only `ResourceHttpContent` needs it (another proposal in this PR)

still needed:

 - remove `HttpContent.release()` (only needed by `CachingHttpContentFactory`, it's too complex and should be rewritten)
 - javadoc `HttpContent`

optional:

 - Replace `ByteBufferPool` with sized `ByteBufferPool` everywhere a pool + size + directness is used (e.g.: `IOResources`)

**Comment by gregw** [comment]
> @gregw I like #12046 but it's not solving where should `ResourceHttpContent.writeTo()` should find its buffer pool.

@lorban if you want to suck #12046 into this PR and close it, I'm OK with that.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPool.java
Comment (gregw) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPool.java]: This is a little better... but I still don't like that wrapping can hide it.

We have adopted a wrapper heavy approach, so marker interfaces do not work well, unless we can unwrap all the layers looking for the marker.

Perhaps we need a Content.Sink.Wrapper class and then a static method that looks through the Sink and any wrappers looking for a Holder?

Comment (gregw) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPool.java]: The other option is to just make getByteBufferPool() a method on both Content.Sink and Content.Source

Comment (lorban) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPool.java]: This issue a way more complex than I anticipated, and since the best is the enemy of the good, I'm giving up and passing the sized pool as a parameter to `writeTo()`.

We could eventually revisit this later on if we find the time, but I don't want to block this PR on this matter.

**Comment by lorban** [comment]
> @lorban if you want to suck #12046 into this PR and close it, I'm OK with that.

I prefer to focus solely on `HttpContent.writeTo()` here as I fear merging the two would turn this PR into an omnibus one. Those two concerns can be worked on separately and in parallel so I've reverted any related changes from this PR.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java
Comment (lorban) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java]: Agreed that this should be taken on in a different PR.

**Comment by lorban** [comment]
@gregw as I respond to one of your comments, the issue of not passing `ByteBufferPool.Sized` to `HttpContent.writeTo()` is way more complex than I anticipated, and since the best is the enemy of the good, I'm giving up and passing the sized pool as a parameter.

I've removed all changes that are not directly related to `HttpContent.writeTo()` so I think this PR is now ready for its review  and to eventually be merged.

**Comment by lorban** [comment]
Latest update: after a conversation with @sbordet, he managed to convince me that the `HttpConfiguration`'s `outputBufferSize` and `useOutputDirectByteBuffers` are unrelated to the buffer size/directness the `ResourceHttpContent` should use.

`ResourceHttpContent` is the only `HttpContent` implementation that requires a sized buffer pool to use in `writeTo()`, other `HttpContent` implementations use buffers with freely decided size/directness.

So I've modified `ResourceHttpContentFactory` so that it takes a sized buffer pool that can then be used by `ResourceHttpContent`, the buffer size and directness being configured by the instantiators of `ResourceHttpContentFactory`, i.e.: `ResourceServlet` or `ResourceService` which allow configuring those parameters via some servlet init parameters and setters respectively.

So the final signature if the new async API finally has this elegant form:
```java
void writeTo(Content.Sink sink, long offset, long length, Callback callback);
```

**Comment by gregw** [comment]
I'm not entirely convinced that the directness is unrelated. But am happy to have an elegant API and use configuration to align sizes and directness.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:121]: Feels weird to have to find the BBP, rather than just passing it in at construction time.

I'd prefer to have a `BBP.Sized` as a parameter to the constructor of `ResourceHandler`, as it is a necessary component.

This would have the benefit of removing the size/direct fields, as they are encapsulated in `BBP.Sized`.

**Review by sbordet — DISMISSED** [review]

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:121]: The way the ServerConnector handles this is to have a full constructor taking a buffer pool, plus one without that passes null.   If null is passed, then then the server is looked at for its buffer pool.

We should have the same logic here.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java
Comment (gregw) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java]: Should this write EOF and then succeed the callback?

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java
Comment (gregw) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java]: Why is this not last?
There are significant efficiencies to be had by sending content with last==true, rather than sending the content and then an empty buffer with last==true.

If you need a writeTo without last==true for multi-part ranges, then I think you need another method taking the boolean (or always pass the boolean).

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java
Comment (lorban) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java]: Yes, I think this should. Let me change that.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java
Comment (lorban) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java]: You just spotted a mistake: the last flag should be set to `true` here (and in the FileMapping impls too).

Multi-part ranges already deal with this by manipulating the last flag, see the comment I added in the [ee9 ResourceService](https://github.com/jetty/jetty.project/blob/219d442b3d4904f4c2f784434558482ba9e394ba/jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ResourceService.java#L868-L873).

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java
Comment (lorban) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:121]: I'm not against adding a constructor that expects a `ByteBufferPool.Sized`, but that would clash with the two properties I added (`_byteBufferSize` and `_useDirectByteBuffers`) that specify the size of the buffers when the pool is taken from the server. There would be some confusion between the size/directness of the properties and the ones of the sized pool.

Or we could add a constructor that takes a non-sized pool, but then the API makes you control the pool both via a constructor argument _and_ a couple properties; that's not very elegant and begs the question _why not passing pool + size + directness as ctor args_ which leads to _why not requiring a sized pool_ which sends us back to square 1.

**Comment by lorban** [comment]
Followup PRs:

 - https://github.com/jetty/jetty.project/pull/12046
 - https://github.com/jetty/jetty.project/pull/12060
 - https://github.com/jetty/jetty.project/pull/12069

Once all those have been merged, the `HttpContent` rework can be considered done.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:121]: I would remove those two properties.  Just have an constructor that that takes a Sized pool.   If that is null then you look for a buffer pool in the server and you can default the buffer size and directness (or look for the first connector and take them from that).

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java]: Remove these and just use an optional constructor injected Sized pool.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java
Comment (gregw) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java]: Maybe move this logic and the length handling below into `InputStreamContentSource`, so it can be used elsewhere.

**Review by gregw — CHANGES_REQUESTED** [review]
Very close now....

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java
Comment (lorban) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:121]: Sounds alright, let me do that.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java
Comment (lorban) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java]: Done.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java
Comment (lorban) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java]: Sure, done.

**Review by gregw — APPROVED** [review]

**Pull request merged**
